File: src/systems/apica.zig
const std = @import("std");
const ApicaRight = @import("../utils/enums.zig").ApicaRight;
const ApicaMode = @import("../utils/enums.zig").ApicaMode;
const bytecode = @import("../common/bytecodes.zig");
const version = @import("../utils/version.zig");

const BytecodeReaderSystem = @import("reader.zig").BytecodeReaderSystem;
const LoggerSystem = @import("logger.zig").LoggerSystem;
const EvaluatorSystem = @import("evaluator.zig").EvaluatorSystem;

pub const ApicaSystem = struct {
    arena: std.heap.ArenaAllocator,
    allocator: std.mem.Allocator,

    actual_mode: ApicaMode,
    actual_right: u8,

    reader: BytecodeReaderSystem,
    logger: LoggerSystem,
    evaluator: EvaluatorSystem,

    pub fn init(main_allocator: std.mem.Allocator) ?*ApicaSystem {
        var self = std.heap.page_allocator.create(ApicaSystem) catch {
            return null;
        };

        self.* = ApicaSystem{
            .arena = std.heap.ArenaAllocator.init(main_allocator),
            .allocator = undefined,
            .actual_mode = ApicaMode.Init,
            .actual_right = @intFromEnum(ApicaRight.MainMenu),

            .reader = undefined,
            .logger = undefined,
            .evaluator = undefined,
        };

        self.allocator = self.arena.allocator();
        self.reader = BytecodeReaderSystem.init(self);
        self.logger = LoggerSystem.init(self, true);

        if (EvaluatorSystem.init(self)) |evaluator| {
            self.evaluator = evaluator;
        } else {
            return null;
        }

        return self;
    }

    pub fn destroy(self: *ApicaSystem) void {
        self.logger.destroy();
        self.reader.destroy();
        self.evaluator.destroy();

        self.arena.deinit();
        std.heap.page_allocator.destroy(self);
    }

    pub fn get_allocator(self: *const ApicaSystem) std.mem.Allocator {
        return self.allocator;
    }

    pub fn get_logger(self: *const ApicaSystem) LoggerSystem {
        return self.logger;
    }

    pub fn is_running(self: *const ApicaSystem) bool {
        return self.actual_mode != ApicaMode.SpecialQuit;
    }

    pub fn quit_app(self: *ApicaSystem) void {
        if (self.actual_right & @intFromEnum(ApicaRight.App_Right) == 0) {
            return;
        }

        self.actual_mode = ApicaMode.Quit;
    }

    pub fn load_app(self: *ApicaSystem, app_name: []const u8) void {
        if (self.actual_right & @intFromEnum(ApicaRight.App_Right) == 0) {
            return;
        }

        if (!self.evaluator.clear_data()) {
            self.actual_mode = ApicaMode.SpecialQuit;
            return;
        }

        if (self.arena.queryCapacity() > 0) {
            _ = self.arena.reset(.retain_capacity);
            self.allocator = self.arena.allocator();
        }

        self.logger.create_file_for(app_name);
        self.reader.read_app(app_name);
    }

    pub fn update_system(self: *ApicaSystem) void {
        switch (self.actual_mode) {
            ApicaMode.Init => {
                const init_node = self.reader.get_entry_node(bytecode.ApicaEntrypointBytecode.Init);
                if (init_node) |r_init| {
                    self.evaluator.evaluate(r_init);
                } else {
                    self.logger.__logn_error(&.{"Failed to load the init entrypoint of the app"});
                }

                self.actual_mode = ApicaMode.Update;
            },

            ApicaMode.Update => {
                const update_node = self.reader.get_entry_node(bytecode.ApicaEntrypointBytecode.Update);
                if (update_node) |r_update| {
                    self.evaluator.evaluate(r_update);
                } else {
                    self.logger.__logn_error(&.{"Failed to load the update entrypoint of the app"});
                    self.actual_mode = ApicaMode.Quit;
                }
            },

            ApicaMode.Quit => {
                const quit_node = self.reader.get_entry_node(bytecode.ApicaEntrypointBytecode.Quit);
                if (quit_node) |r_quit| {
                    self.evaluator.evaluate(r_quit);
                } else {
                    self.logger.__logn_error(&.{"Failed to load the quit entrypoint of the app"});
                }

                switch (self.actual_right) {
                    @intFromEnum(ApicaRight.MainMenu) => self.actual_mode = ApicaMode.SpecialQuit,
                    @intFromEnum(ApicaRight.App) => {
                        self.actual_mode = ApicaMode.Init;
                        self.actual_right = @intFromEnum(ApicaRight.MainMenu);
                        self.load_app("APICA_MENU");
                    },

                    else => {},
                }
            },

            ApicaMode.SpecialQuit => {},
        }
    }

    pub fn get_apica_version_major(_: *ApicaSystem) u64 {
        return version.VERSION_MAJOR;
    }

    pub fn get_apica_version_minor(_: *ApicaSystem) u64 {
        return version.VERSION_MINOR;
    }
};

File: src/systems/apica.rs
use apica_common::bytecodes::ApicaEntrypointBytecode;
use apica_common::values::value::Value;
use winit::application::ApplicationHandler;
use winit::event::WindowEvent;
use winit::event_loop::ActiveEventLoop;
use winit::window::{Window, WindowId};
use crate::systems::evaluator::EvaluatorSystem;
use crate::systems::inputs::InputsSystem;
use crate::systems::logger::LoggerSystem;
use crate::systems::reader::BytecodeReaderSystem;
use crate::systems::rights::RightSystem;
use crate::systems::window::WindowSystem;
use crate::utils::rights::{ApicaMode, ApicaRight, APICA_MAIN_MENU};

pub struct ApicaSystem {
    rights: RightSystem,
    reader: BytecodeReaderSystem,
    logger: LoggerSystem,
    evaluator: EvaluatorSystem,
    window: WindowSystem,
    inputs: InputsSystem,
}

impl ApicaSystem {
    pub fn init() -> ApicaSystem {
        ApicaSystem{
            rights: RightSystem::init(),
            reader: BytecodeReaderSystem::init(),
            logger: LoggerSystem::init(true),
            evaluator: EvaluatorSystem::init(),
            window: WindowSystem::init(),
            inputs: InputsSystem::init(),
        }
    }

    pub fn is_running(&self) -> bool {
        self.rights.is_running()
    }

    pub fn quit_app(&mut self) {
        self.rights.quit_app();
    }

    pub fn load_app(&mut self, app_name: &str) {
        if !self.rights.has_right(ApicaRight::AppRight) {
            return;
        }

        self.evaluator.clear_data();
        self.logger.create_file_for(app_name);
        self.reader.read_app(app_name, &mut self.logger);

        let title = if let Some(value) = self.reader.get_data("title") && let Value::String(title) = value {
            title.get_value().as_ref().unwrap()
        } else {
            "???"
        };
        self.window.system_set_title(title);
    }

    pub fn update_system(&mut self) {
        match self.rights.get_mode() {
            ApicaMode::SpecialQuit => {},

            ApicaMode::SpecialInit => {
                self.rights.set_mode(ApicaMode::Init);
                self.load_app(APICA_MAIN_MENU)
            },

            ApicaMode::Init => {
                if let Some(init_node) = self.reader.get_entry_node(ApicaEntrypointBytecode::Init) {
                    self.evaluator.evaluate(init_node, &mut self.logger, &mut self.rights, &mut self.window, &mut self.inputs);
                } else {
                    self.logger.system_logn_error("Failed to load the init entrypoint of the app".to_string());
                }

                self.rights.set_mode(ApicaMode::Update);
            },

            ApicaMode::Update => {
                if let Some(update_node) = self.reader.get_entry_node(ApicaEntrypointBytecode::Update) {
                    self.evaluator.evaluate(update_node, &mut self.logger, &mut self.rights, &mut self.window, &mut self.inputs);
                } else {
                    self.rights.set_mode(ApicaMode::Quit);
                    self.logger.system_logn_error("Failed to load the update entrypoint of the app".to_string());
                }
            },

            ApicaMode::Quit => {
                if let Some(quit_node) = self.reader.get_entry_node(ApicaEntrypointBytecode::Quit) {
                    self.evaluator.evaluate(quit_node, &mut self.logger, &mut self.rights, &mut self.window, &mut self.inputs);
                } else {
                    self.logger.system_logn_error("Failed to load the quit entrypoint of the app".to_string());
                }

                if self.rights.has_right(ApicaRight::MainMenuRight) {
                    self.rights.set_mode(ApicaMode::SpecialQuit);
                } else if self.rights.has_right(ApicaRight::AppRight) {
                    self.rights.set_mode(ApicaMode::Init);
                    self.rights.set_right(ApicaRight::MainMenu);
                    self.load_app(APICA_MAIN_MENU);
                }
            },
        }
    }
}

impl ApplicationHandler for ApicaSystem {
    fn resumed(&mut self, event_loop: &ActiveEventLoop) {
        let attributes = Window::default_attributes()
            .with_title("Apica");

        let window = event_loop.create_window(attributes).unwrap();
        self.window.set_window(window);
    }

    fn window_event(&mut self, _event_loop: &ActiveEventLoop, _window_id: WindowId, event: WindowEvent) {
        match event {
            WindowEvent::CloseRequested => {
                self.rights.quit_app();
            },

            WindowEvent::KeyboardInput { event, .. } => {
                self.inputs.handle_key_event(&event);
            },

            _ => {},
        }
    }

    fn about_to_wait(&mut self, event_loop: &ActiveEventLoop) {
        self.update_system();
        if !self.is_running() {
            event_loop.exit();
        }
    }
}